File: server_nodejs/src/repositories/profile-repository.ts
import {getDatabase} from "./base";
import {Profile} from "../models/profile";
import {AddProfileArgs} from "../interfaces/profile-interfaces";
import {logger} from "../utils/log-utils";

const INSERT_PROFILE = `INSERT INTO profiles (user_id, bio) VALUES (@userId, @bio)`;
const UPDATE_PROFILE = `UPDATE profiles SET bio = @bio, updated_at = @updatedAt WHERE user_id = @userId`;
const SELECT_PROFILE = `
SELECT 
    id, 
    bio, 
    user_id as userId,
    strftime('%s', datetime(created_at, 'localtime')) as createdAt,     
    strftime('%s', datetime(updated_at, 'localtime')) as updatedAt
    FROM profiles 
WHERE user_id = @userId`;

export const profileInsert = (data: AddProfileArgs): Profile => {
    logger.verbose(`Inserting profile: ${data}`);
    const db = getDatabase();

    const insert = db.prepare(INSERT_PROFILE);
    insert.run(data);
    return profileGet(data.userId);
};

export const profileUpdate = (data: AddProfileArgs): Profile => {
    logger.verbose(`Updating profile for userId:: ${data.userId}`);
    const db = getDatabase();
    const update = db.prepare(UPDATE_PROFILE);
    const updatePayload = {
        ...data,
        updatedAt: new Date().toISOString()
    }
    update.run(updatePayload);

    return profileGet(data.userId);
};

export const profileGet = (userId: number): Profile => {
    logger.verbose(`Fetching profile for user Id: ${userId}`);
    const db = getDatabase();
    const select = db.prepare(SELECT_PROFILE);
    const profile = select.get({userId});

    profile.createdAt = new Date((+profile.createdAt)*1000);
    profile.updatedAt = new Date((+profile.updatedAt)*1000);

    return profile as Profile;
}

File: server_python/src/repositories/profile_repository.py
# -*- coding: utf-8 -*-
from src.models.inputs.add_profile_args import AddProfileArgs
from src.models.profile import Profile
from src.repositories.database import get_sqlite_conn, parse_row
from src.settings import LOGGER

INSERT_PROFILE = "INSERT INTO profiles (user_id, bio) VALUES (:user_id, :bio)"
UPDATE_PROFILE = "UPDATE profiles SET bio = :bio, updated_at = :updated_at WHERE user_id = :user_id"
SELECT_PROFILE = """
SELECT 
    id, 
    bio, 
    user_id as user_id,
    datetime(created_at, 'localtime') as created_at,     
    datetime(updated_at, 'localtime') as updated_at
    FROM profiles 
WHERE user_id = :user_id"""


def profile_get(user_id: int) -> Profile:
    LOGGER.info(f"Fetching profile for user Id: {user_id}")
    db = get_sqlite_conn()
    cursor = db.cursor()
    cursor.execute(SELECT_PROFILE, {"user_id": user_id})
    rows = cursor.fetchall()
    if len(rows) == 0:
        raise ValueError(f"No profile found profile for id {user_id}")
    profile: Profile = parse_row(rows[0], cursor.description, Profile)
    cursor.close()
    return profile


def profile_insert(data: AddProfileArgs) -> Profile:
    LOGGER.info(f"Inserting profile: {data}")
    db = get_sqlite_conn()
    cursor = db.cursor()
    cursor.execute(INSERT_PROFILE, data.__dict__)
    db.commit()
    cursor.close()
    return profile_get(data.user_id)


def profile_update(data: AddProfileArgs) -> Profile:
    LOGGER.info(f"Updating profile for userId: {data.user_id}")
    db = get_sqlite_conn()
    cursor = db.cursor()
    cursor.execute(UPDATE_PROFILE, data.__dict__)
    db.commit()
    cursor.close()
    return profile_get(data.user_id)
